### ISSUE TITLE
Request: Add window cycling with scroll wheel,  just like it is implemented in the default window-list@cinnamon setting ""Enable mouse-wheel scrolling in the window list"

### ISSUE THREAD

**Issue by gbit00** [klangman/CassiaWindowList#23]
Title kind of says it all, but basically, I would love to be able to scroll through open apps and windows, in order, with the  scroll wheel while hovering over window list.  Would love it even more if that was possible to implement over the entirety of panel (window list & blank space). The extension smart-panel@mohammad-sn can unlock a similar feature, but only on blank space, the cycling order is not based on the order of apps in the window list, and all of the jumping around drives me crazy!

 I have also finally replied back to your questions regarding my previous and now-closed issue involiving icon-label spacing.

GREAT JOB AGAIN BTW this project of yours has kept my sanity and ability to keep using linux and in turn has deepened my appreciation of operating systems and computers in general.  THANK YOU!!

**Comment by klangman** [comment]
Yes, I think I can implement the option to scroll through the windows using the scroll-wheel while hovering the window-list, but it might be more difficult to do that for the empty sections of the panel. I will keep in it mind and see what I can do without doing too much heroics. 

I'll check-out that feature in the default window-list and see how it behaves. I have to think about how it should work.. i.e. should it start scrolling based on the button being hovered over, or start using the current focused window.. In any case the feature will be get implemented.

Thanks for the words of encouragement! It nice to know that my work is appreciated and that it's getting used by so many people.

**Comment by gbit00** [comment]
I just checked it out, and like I imagined, probably for ease-of-use purposes, it scrolls through windows based on the current focused window, completely regardless of what button is being hovered over.  That way, users don't have to think much, just scroll over your window list and the windows change focus. Easy-peasy from a user perspective, hopefully doesn't create too much headache on your end!  Obviously, this setting would have to take precedence/ignore the individual scroll up and down functions you have already implemented.  (On that note, maybe it's just me, but there does seem to be some buginess in getting your scroll wheel functions to activate...  I have found that by coming in to the button from the side while already scrolling is the most reliable way, but that is obviously not ideal).  Cheers!

**Comment by gbit00** [comment]
And as far as the empty sections bit, the (possibly!) easiest strategy that first comes to mind would be for the applist to (invisibly) expand to the entire length of the panel, while respecting a small space at either end to allow for other applets to be added.  But I know next to nothing about coding so what sounds easy might not be at all.  Anyway, thanks for listenin'!

**Comment by klangman** [comment]
Expanding the applets space allocation would cause other side effects that are not so pleasant. Trying to intercept the scroll-wheel event on the panel might interfere with smart-panel for those that use it. So it might be tricky to get it working just right in all cases. In any case, I can easily do the work to make the scroll-wheel work like you want when scrolling on the applet so I will do that first (already got it mostly working) and think about the empty space thing after that.

**Comment by gbit00** [comment]
That's awesome! I have set up the installing of your applet directly from github and am eagerly awaiting your developments!  But no rush, hehe, it's just a window list.  Cheers!

**Comment by klangman** [comment]
Just pushed some changes with this feature. Let me know if you find any issues with the new options.

**Comment by gbit00** [comment]
Awesome!  Took me a while to get it working...  all manners of preference backups were getting in the way but starting removing all traces of the old cassia and starting from scratch did the trick!
Just one thing...
The scroll wheel functions are going the wrong way 😂 (at least compared to main-window list and my personal preference).  I have noticed, in general, there does soom to be some huge divide on which way things should go with the scroll wheel, but to me, scrolling down, should make scroll bars go down (on pages, folders, etc), things move towards the right (on seek bars etc), and things getting smaller (for zooming).  Im not sure where or when the divide happens, maybe something to do with touch pads, but yeah, I have normal zoom directions in my mouse configs, and would normally expect a scroll down to move to the next (right) window, while the current behavior is just the opposite.  If you try it on main-indow list you will (should?) see what I mean.  Thoughts?

**Comment by klangman** [comment]
Yep.. It's backwards (but it is what my preferences would be). I'll make it consistent with how the grouped window list works (and how chrome works).

**Comment by klangman** [comment]
OK... I push a change to reverse the direction.

**Comment by klangman** [comment]
I pushed the changes to cinnamon spices.. Version 2.3.7 has the changes.

**Comment by klangman** [comment]
@gbit00 Did you have a chance to test out V2.3.7. Does it work the way you hoped for?

**Comment by gbit00** [comment]
Yes!. 10/10 A+++!   I was wondering, where did you get your inspiration to build-a-better-taskbar?  Have you had any experience with the Windows 10 program 7+ Taskbar Tweaker? https://tweaker.ramensoftware.com/.  If you haven't already, and have a windows 10 machine or VM handy  to play around with, it may be worth checking out. Lots of useful and interesting approaches, ideas, options and functionality. 
Cheers and thanks again... the scroll wheel functionality works like a charm!

**Comment by klangman** [comment]
I started down this path because my work has me using 5 terminal windows (I don't like using tabs because I need to see 2 terminal windows side by side frequently). I decided it would be nice to have all the term windows maintained in a "pool" on the window-list. I also wanted to have only one label for the set of term windows in the "pool" so I could save some window-list real-estate. So I started looking into how that could be done and before you know it I had the features I wanted and a ton of other ideas popping up in my head for how I could improve the window-list to help my efficiency. 

I haven't used Windows for my work or home machines since about 2006. So, no, I didn't see "7+ Taskbar Tweaker" before starting this work, but I do recall the name, so someone must have pointed me at it before now. Currently I don't really have access to Windows machines other than a few VMs at work that I am not really free to mess with.

I am glad you like the my applet. It took a lot of work to get it to where it is now, so it would be a shame if I was the only one using it. At this point the efficiency improvements it brings to my daily work will never make up for the time I spent working on coding it, so if enough others start using it, one day it might have saved enough time to cover the time I spent developing it :-)

I did learn a lot working on this project, and I was able to use that knowledge on a bunch of other applets and extensions, so it was time well spent I guess.

**Comment by gbit00** [comment]
Absolutely, time well spent.  Life is more  of a journey, right?  It seems like if one must put in the hard work to learn what you did, the most fulfilling way to learn those lessons has to be when working with YOUR own project/dream/program. The crazy part is, we never really know to the full extent just how much our creations and actions really do affect the wide world.
The fact that humans and most especially engineers will often spend way more effort than is "rational" on making perhaps a simple task just a little bit simpler, and sharing that with the world, I find incredibly charming.  The entire existence of computing, mass production, and even electricity, was perhaps born from this primordial urge to put brain effort towards simplifying body-effort problems.  The fact that stellar individuals have donated and shared their innovations for all to profit from is uniquely human and amazing.  So thanks again for being a contributor to the open-source-repository of knowledge and tools.  I know your applet makes the entire linux universe possible/palatable to a taskbar junkie like myself.  Cheers!

**Comment by klangman** [comment]
Thanks again for the words of encouragement. 

I hope to resume working on your icon-spacing issue soon, I have a few other things to finish before I get back to it but it's on my to-do list.

I think this issue is now taken care of? We can close it out now?

**Comment by gbit00** [comment]
👍
